kerastuner/engine/oracle.py

```diff
@@ -17,90 +17,292 @@
+from kerastuner.engine import metrics_tracking
+from kerastuner.engine import stateful
+from kerastuner.engine import trial as trial_lib
+from ..abstractions.tensorflow import TENSORFLOW_UTILS as tf_utils
 
-class Oracle(object):
 
-    def __init__(self):
-        self.space = []
+Objective = collections.namedtuple('Objective', 'name direction')
 
-    def update_space(self, new_entries):
-        """Add new hyperparameters to the tracking space.
 
-        Already recorded parameters get ignored.
+class Oracle(stateful.Stateful):
+    """Implements a hyperparameter optimization algorithm.
 
-        Args:
-            new_entries: A list of HyperParameter objects to track.
-        """
-        ref_names = {p.name for p in self.space}
-        for p in new_entries:
-            if p.name not in ref_names:
-                self.space.append(p)
+    Attributes:
+        objective: String. Name of model metric to minimize
+            or maximize, e.g. "val_accuracy".
+        max_trials: The maximum number of hyperparameter
+            combinations to try.
+        hyperparameters: HyperParameters class instance.
+            Can be used to override (or register in advance)
+            hyperparamters in the search space.
+        tune_new_entries: Whether hyperparameter entries
+            that are requested by the hypermodel
+            but that were not specified in `hyperparameters`
+            should be added to the search space, or not.
+            If not, then the default value for these parameters
+            will be used.
+        allow_new_entries: Whether the hypermodel is allowed
+            to request hyperparameter entries not listed in
+            `hyperparameters`.
+    """
+
+    def __init__(self,
+                 objective,
+                 max_trials=None,
+                 hyperparameters=None,
+                 allow_new_entries=True,
+                 tune_new_entries=True):
+        self.objective = _format_objective(objective)
+        self.max_trials = max_trials
+        if not hyperparameters:
+            if not tune_new_entries:
+                raise ValueError(
+                    'If you set `tune_new_entries=False`, you must'
+                    'specify the search space via the '
+                    '`hyperparameters` argument.')
+            if not allow_new_entries:
+                raise ValueError(
+                    'If you set `allow_new_entries=False`, you must'
+                    'specify the search space via the '
+                    '`hyperparameters` argument.')
+            self.hyperparameters = hp_module.HyperParameters()
+        else:
+            self.hyperparameters = hyperparameters
+        self.allow_new_entries = allow_new_entries
+        self.tune_new_entries = tune_new_entries
+
+        # trial_id -> Trial
+        self.trials = {}
+        # tuner_id -> Trial
+        self.ongoing_trials = {}
+
+    def _populate_space(self, trial_id):
```

Comment:
Nit: private method (this one and below) should be grouped together at the end, to make the code more readable (public method first, and more commonly used methods first).

Reply:
I wanted to put this and _score_trial first since they are the most likely methods for Oracle subclasses to override. But I could put it below the public methods, WDYT?